lighthouse/mcs-compliance.md

```diff
@@ -0,0 +1,176 @@
+PodIPs in `EndpointSlices`.
+
+## Proposal
+
+When creating `ServiceExport`, users will have option to add annotation for enabling VIP for that Service and another
+to use PodIPs in `EndpointSlices`. These annotations will also be added to the aggregated `ServiceImport`.
+
+Option to add PodIPs to EndpointSlice will only be applicable when VIP option is also enabled. A use case that requires
+PodIPs but not VIP can already be taken care of by using HeadlessServices.
+
+Configuration flags will also be added to `Submariner` and `ServiceDiscovery` CRDs, and `subctl` to set these options at
+global level. This is for deployments where user only wants to use one specific option and doesn't want to set the
+annotation explicitly for each `ServiceExport`.
+
+Default behavior with nothing configured would be same as it is currently i.e. no VIP and no PodIPs.
+
+Global flag can't be changed at runtime and requires a reinstall of whole clusterset. Change of flag in CRs will require
+a reinstall of Submariner on that cluster as it changes he behavior of `ServiceExports` without annotation.
+
+### VIP CIDR and allocation
+
+1. Default CIDR of `243.0.0/8` will be used for Clusterset. User can change this at the time of deploying broker.
+2. Each Cluster will get a default subset CIDR of `243.0.0.0/16` for allocating VIPs to Services exported from that
+cluster. User can change this at the time of join but this CIDR must be an unallocated subset of global VIP CIDR.
+3. VIP allocation is done by the first cluster to export the `Service`.
+4. When Service is exported on subsequent clusters, no VIP allocation is done if VIP already present on `ServiceImport`.
+5. VIP is deallocated only when `ServiceImport` is deleted i.e. `Service` or `ServiceExport` are no longer present on any
+of the clusters.
+6. If Submariner is uninstalled on the cluster that allocated a given VIP, VIP is not changed on `ServiceImport`.
+7. When `lighthouse-agent` starts, it should check `ServiceImport`s for any VIPs allocated from its VIP CIDR. Any such
+VIPs should be added to its local VIP allocation cache. This will also help with scenarios where cluster that allocated
+VIP has uninstalled Submariner and another cluster got the same VIP CIDR on installation.
+
+Note that `1` and `2` above are similar to Globalnet CIDR.
+
+### DNS Resolution
+
+Lighthouse DNS will return VIP for `ServiceImports` if available. It will be up to any other solution external to
+Submariner to load balance this VIP to PodIPs in `EndpointSlices`. If VIP is not present, it will behave as it does
+today, i.e. return one of the ClusterIPs from `EndpointSlices`.
+
+### Conflict Resolution
+
+In case of conflict in VIP and PodIP settings when exporting `Service`, configuration on the oldest `ServiceExport`
+will be used. The `ServiceExport` in conflict will not be exported and `ServiceExportConflict` condition will be set.
```

Comment:
We could also resolve the conflict by merging, ie any `ServiceExport` with the annotation results in a VIP. The first exporting cluster with the annotation allocates the VIP.

Reply:
> The first exporting cluster with the annotation allocates the VIP.

I should clarify - that's if the annotation is set to `true`. We could still have a conflict if another cluster explicitly sets their annotation to `false`. In that case, report the conflict but take the `true` setting.

Another interesting scenario - what if the global VIP option is set but a `ServiceExport's` annotation is set to `false`.... this would be a way to disable VIP for a particular service.

Reply:
Thinking more on this, I think we should keep this as simple as possible, ie VIP enablement for a service is determined by the exporting cluster that creates the aggregated `ServiceImport`.  So:

  - if global flag is set AND either `ServiceExport` annotation is not present OR the annotation is set to true THEN VIP is enabled

  - if global flag is not set OR the annotation is present AND set to true THEN VIP is enabled

Reply:
Just to be clear, regardless of whether not the user specifies the global VIP flag, we always need to allocate a VIP CIDR for each cluster from the clusterset CIDR to support the ServiceExport annotation.

Reply:
> > The first exporting cluster with the annotation allocates the VIP.
> 
> I should clarify - that's if the annotation is set to `true`. We could still have a conflict if another cluster explicitly sets their annotation to `false`. In that case, report the conflict but take the `true` setting.
> 
Yes. I didn't word it properly. Maybe hsould use the phrase "if VIP is enabled" and VIP enabled definition is the combination of global+annotation as you clarified.
> Another interesting scenario - what if the global VIP option is set but a `ServiceExport's` annotation is set to `false`.... this would be a way to disable VIP for a particular service.
Yes. Global flag is only used when annotaiton is not available on export. If available, it overrides whatever is set in Global.